CMakeLists.txt

```diff
@@ -48,108 +60,167 @@ include_directories(./include)
+endif()
 
-set(PROGRAMS_J1939_TIMEDATE
-  j1939-timedate-srv
-  j1939-timedate-cli
-)
+if(ENABLE_MEASUREMENT)
+  list(APPEND PROGRAMS
+    can-calc-bit-timing
+    canfdtest
+    canerrsim
+  )
+  list(APPEND PROGRAMS_CANLIB canbusload)
+endif()
 
-set(PROGRAMS_J1939_VEHICLE_POSITION
-  j1939-vehicle-position-srv
-)
+if(ENABLE_ISOTP)
+  list(APPEND PROGRAMS
+    isotpdump
+    isotpperf
+    isotprecv
+    isotpsend
+    isotpsniffer
+    isotptun
+  )
 
-set(PROGRAMS_ISOBUSFS
-  isobusfs-srv
-  isobusfs-cli
-)
+  if(HAVE_FORK)
+    list(APPEND PROGRAMS isotpserver)
+  endif()
+endif()
 
-set(PROGRAMS
-  ${PROGRAMS_CANLIB}
-  canerrsim
-  canfdtest
-  cangw
-  cansniffer
-  isotpdump
-  isotpperf
-  isotprecv
-  isotpsend
-  isotpsniffer
-  isotptun
-  slcan_attach
-  slcand
-)
+if(ENABLE_J1939)
+  list(APPEND PROGRAMS_J1939
+    j1939acd
+    j1939cat
+    j1939spy
+    j1939sr
+    testj1939
+    j1939-timedate-srv
+    j1939-timedate-cli
+  )
+  if(ENABLE_GPS)
+    list(APPEND PROGRAMS_J1939 j1939-vehicle-position-srv)
+  endif()
+endif()
 
-if(HAVE_FORK)
-  list(APPEND PROGRAMS bcmserver)
-  list(APPEND PROGRAMS isotpserver)
+if(ENABLE_ISOBUSFS)
+  list(APPEND PROGRAMS_ISOBUSFS
+    isobusfs-srv
+    isobusfs-cli
+  )
 endif()
 
-add_executable(can-calc-bit-timing
-  calc-bit-timing/can-calc-bit-timing.c
-)
+if(ENABLE_LOG_FILE)
+  list(APPEND PROGRAMS_CANLIB
+    asc2log
+    log2asc
+    log2long
+  )
+endif()
+
+if(ENABLE_SLCAN)
+  list(APPEND PROGRAMS
+    slcan_attach
+    slcand
+  )
+  list(APPEND PROGRAMS_CANLIB slcanpty)
+endif()
+
+if(ENABLE_MCP251XFD)
+  list(APPEND PROGRAMS mcp251xfd-dump)
+endif()
 
-add_executable(mcp251xfd-dump
-  mcp251xfd/mcp251xfd-dev-coredump.c
-  mcp251xfd/mcp251xfd-dump.c
-  mcp251xfd/mcp251xfd-main.c
-  mcp251xfd/mcp251xfd-regmap.c
+list(APPEND PROGRAMS
+  ${PROGRAMS_CANLIB}
+  ${PROGRAMS_J1939}
+  ${PROGRAMS_ISOBUSFS}
 )
 
-if(NOT ANDROID)
-  list(APPEND PROGRAMS ${PROGRAMS_J1939})
+add_library(can STATIC EXCLUDE_FROM_ALL
+  lib.c
+  canframelen.c
+)
 
-  add_library(j1939 STATIC
+if(ENABLE_J1939)
+  add_library(j1939 STATIC EXCLUDE_FROM_ALL
     libj1939.c
   )
-
   target_link_libraries(j1939
     PRIVATE can
   )
+endif()
 
-  add_library(isobusfs
+if(ENABLE_ISOBUSFS)
+  add_library(isobusfs EXCLUDE_FROM_ALL
     isobusfs/isobusfs_cmn.c
     isobusfs/isobusfs_cmn_dh.c
   )
-
   set(PUBLIC_HEADER_ISOBUSFS
     isobusfs/isobusfs_cmn.h
     isobusfs/isobusfs_cmn_cm.h
   )
-
   set_target_properties(isobusfs PROPERTIES
     PUBLIC_HEADER "${PUBLIC_HEADER_ISOBUSFS}"
     SOVERSION 0
   )
-
   install(TARGETS isobusfs
     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
     PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
   )
+endif()
+
+if(can-calc-bit-timing IN_LIST PROGRAMS)
+  list(REMOVE_ITEM PROGRAMS can-calc-bit-timing)
+  add_executable(can-calc-bit-timing
+    calc-bit-timing/can-calc-bit-timing.c
+  )
+  install(TARGETS can-calc-bit-timing DESTINATION ${CMAKE_INSTALL_BINDIR})
+endif()
```

Comment:
Can you move the:

```
  list(REMOVE_ITEM PROGRAMS can-calc-bit-timing)
  add_executable(can-calc-bit-timing
    calc-bit-timing/can-calc-bit-timing.c
  )
  install(TARGETS can-calc-bit-timing DESTINATION ${CMAKE_INSTALL_BINDIR})
```

to the place where `can-calc-bit-timing` is added to the `PROGRAMS` in the first place? Same for the other `list(REMOVE_ITEM ...)`?

Reply:
Of course. I was hoping to create a bit of separation between the "program selection" and the actual build part, but I can see where you're coming from.